rust/src/delta.rs

```diff
@@ -1355,6 +1455,223 @@ fn log_entry_from_actions(actions: &[Action]) -> Result<String, serde_json::Erro
+        #[from]
+        source: ArrowError,
+    },
+    /// Passthrough error returned when calling StorageBackend.
+    #[error("StorageError: {source}")]
+    Storage {
+        /// The source StorageError.
+        #[from]
+        source: StorageError,
+    },
+    /// Passthrough error returned by serde_json.
+    #[error("serde_json::Error: {source}")]
+    JSONSerialization {
+        /// The source serde_json::Error.
+        #[from]
+        source: serde_json::Error,
+    },
+}
+
+/// Struct for writing checkpoints to the delta log.
+pub struct CheckPointWriter {
+    table_uri: String,
+    delta_log_path: String,
+    last_checkpoint_path: String,
+    storage: Box<dyn StorageBackend>,
+    schema_factory: DeltaLogSchemaFactory,
+}
+
+impl CheckPointWriter {
+    /// Creates a new CheckPointWriter.
+    pub fn new(table_uri: &str, storage: Box<dyn StorageBackend>) -> Self {
+        let delta_log_path = storage.join_path(table_uri, "_delta_log");
+        let last_checkpoint_path = storage.join_path(delta_log_path.as_str(), "_last_checkpoint");
+        let schema_factory = DeltaLogSchemaFactory::new();
+
+        Self {
+            table_uri: table_uri.to_string(),
+            delta_log_path,
+            last_checkpoint_path,
+            storage,
+            schema_factory,
+        }
+    }
+
+    /// Creates a new checkpoint at the specified version.
+    /// NOTE: This method loads a new instance of delta table to determine the state to
+    /// checkpoint.
+    pub async fn create_checkpoint_for_version(
+        &self,
+        version: DeltaDataTypeVersion,
+    ) -> Result<(), CheckPointWriterError> {
+        let table = open_table_with_version(self.table_uri.as_str(), version).await?;
+
+        self.create_checkpoint_from_state(version, table.get_state())
+            .await
+    }
+
+    /// Creates a new checkpoint at the specified version from the given DeltaTableState.
+    pub async fn create_checkpoint_from_state(
+        &self,
+        version: DeltaDataTypeVersion,
+        state: &DeltaTableState,
+    ) -> Result<(), CheckPointWriterError> {
+        // TODO: checkpoints _can_ be multi-part... haven't actually found a good reference for
+        // an appropriate split point yet though so only writing a single part currently.
+
+        info!("Writing parquet bytes to checkpoint buffer.");
+        let parquet_bytes = self.parquet_bytes_from_state(state)?;
+
+        let size = parquet_bytes.len() as i64;
+
+        let checkpoint = CheckPoint {
+            version,
+            size,
+            parts: None,
+        };
+
+        let file_name = format!("{:020}.checkpoint.parquet", version);
+        let checkpoint_path = self.storage.join_path(&self.delta_log_path, &file_name);
+
+        info!("Writing checkpoint to {:?}.", checkpoint_path);
+        self.storage
+            .put_obj(&checkpoint_path, &parquet_bytes)
```

Comment:
for naming consistency sake, since https://github.com/delta-io/delta-rs/pull/258#issuecomment-850925010, we use uri to refer to object reference that contains both scheme,  host and path, for example `s3://bucket/key`. path suffix is used for variables that only hosts the path part. For example `key`.

So I think a better name here would be `checkpoint_uri`.

Reply:
fixed. also noticed a couple other uris i was calling paths and fixed those.